synapse/config/registration.py

```diff
@@ -128,6 +128,22 @@ def read_config(self, config, **kwargs):
             int
         ] = refreshable_access_token_lifetime
 
+        # The `nonrefreshable_access_token_lifetime` applies for tokens that can NOT be
+        # refreshed using a refresh token.
+        # If it is None, then these tokens last for the entire length of the session,
+        # which is infinite by default.
+        # The intention behind this configuration option is to help with requiring
+        # all clients to use refresh tokens, if the homeserver administrator requires.
```

Comment:
You mentioned this in the docs PR but I can't remember the reasoning.

For the use case you describe, it sounds like a server admin would want a boolean flag which, if set:
- will make Synapse invalidate all non-refreshable access tokens currently in its DB on startup
- will make Synapse refuse to grant any new non-refreshable tokens

Or perhaps the point is that I can retroactively enforce a lifetime to tokens I've already granted, so that after $CUTOFF everyone will have to use refreshable tokens?

Reply:
Not caring about previously-issued tokens is a simplication that all the current code takes advantage of (even `session_lifetime` and friends do this).
If that worries us we could do something about it in the future but probably not relevant right now.

This option is only relevant to newly-minted non-refreshable access tokens.

Setting a short lifetime on non-refreshable access tokens is to satisfy the use case that ALL client sessions that go inactive for 15 minutes (or whatever) must be logged out.
Before this change, clients not supporting refresh tokens will be given an access token for the lifetime of `session_lifetime` (typically a long time).
After this change, that can be narrowed down further by this additional setting. For example, it could be set to `15 minutes` to give a similar level of security as refreshable tokens.
Of course, UX will suffer — so you'd want to be sure that all your users are using clients that support refresh tokens, but the goal is to close the gap that allows bypassing the auto-logout mechanism (or in other words: prevents issuing long-lived access tokens).

Reply:
> Not caring about previously-issued tokens is a simplication that all the current code takes advantage of [...]

Okay, but ignoring existing tokens seems at odds with the use case of "requiring all clients to use refresh tokens".  That sounds like a pretty cut-and-dry security requirement which I imagine e.g. an admin operating a security-conscious company homeserver would want to enforce universally, even to existing clients.

Perhaps you're aiming to also facilitate an admin who who wants to encourage, but not require refresh tokens.

In a nutshell, I'm not sure what you mean by `requiring` in the quoted source code.

[more to follow]

Reply:
> Setting a short lifetime on non-refreshable access tokens is to satisfy the use case that ALL client sessions that go inactive for 15 minutes (or whatever) must be logged out.

Okay, so this is a best-effort means to continue service to old clients while still enforcing the activity requirement.

If we're going to ignore existing non-refreshable access tokens though, we can't say that we've enforced this requirement for all clients. (Perhaps you want to make further changes to that end in an upcoming PR?)

Reply:
I guess you're making a fair point. Perhaps the answer here is that the admin needs to revoke all the existing tokens first (or, of course, set up the server using this facility from day one).
It doesn't really seem any different from the security requirement that `session_lifetime` imposes though and that is also something that is determined when an access token is issued (and not updated retrospectively).

Perhaps this is something that should be addressed through documentation? The current `session_lifetime` config option makes note of this fact, perhaps the new options need a similar note in the sample config?

Reply:
Reflecting on this a bit, I don't think I had any objections to the proposed source code changes. I can also see that everything you've said above corresponds faithfully to the MSC.

I think I'd be happy with
- notes in the docs and config which explains that the settings don't apply retroactively
- an explanation in the docs of how to nuke existing access tokens if admins really want to kill off existing non-refreshable tokens
- a guide/FAQ in the docs of how to enforce the things we're talking about here
  - requiring people to use refresh tokens
  - requiring people to refresh once every $THRESHOLD

Reply:
Taking these suggestions to the other PRs (the config notices are already in #11435 now(